woocommerce/woocommerce-ios
# [Mobile Payments] [SDK 2] Firmware updates for StripeTerminal 2.0

This PR implements the new software update UI for required updates. Note that this targets the `feature/stripe-sdk-2` branch.
With Stripe Terminal 2.0, these are handled a bit differently than optional updates, so I'll handle those in a separate PR.

0%|53%|Complete
-|-|-
![stripe2-required-update-0](https://user-images.githubusercontent.com/8739/134150357-c7cf11ac-b3c6-4902-bddc-a33a3cea7ae4.png)|![stripe2-required-update-53](https://user-images.githubusercontent.com/8739/134150365-7de573e8-476a-41d2-b74a-736b14e7e1b5.png)|![stripe2-required-update-complete](https://user-images.githubusercontent.com/8739/134150369-d1962dbf-4424-4440-a5a9-011a03e1de77.png)

Some notes on implementation:

- `CardReaderSoftwareUpdateState`. Previously the service published a `Float` value with the installation progress, because the rest of the information came implicit from the user initiating the update. Now the UI layer can observe this publisher to have a more complete view of the state regarding software updates.
- `FallibleCancelable`. I went back and forth with this one a few times. I needed a way to pass the "cancellable" that Stripe gives us when an update starts so the UI can have a "Cancel" button. I didn't want to expose the type from Stripe directly, and I thought of using Combine's `Cancellable`, but this is different because the cancel operation might fail with an error. I added a new simple type to represent this, and named it `FallibleCancelable` to avoid confusion with Combine, and because it represents what it does more precisely.
- Progress animation. This might not be the most elegant way to solve this, but it works well enough for something that will be shown infrequently. I would have preferred to draw directly on the screen, but `CardPresentPaymentsModalViewModel` wants a `UIImage` and I didn't want to start refactoring all the modals for this. 

## To test

1. On Xcode, edit the WooCommerce scheme, and enable `-simulate-stripe-card-reader` in the Arguments tab
2. In `StripeCardReaderService.start(_:)`, after the SDK has been initialized (I add this right after `Terminal.shared.delegate = self`), add the following line: `Terminal.shared.simulatorConfiguration.availableReaderUpdate = .required`
3. Run the app
4. Go to Settings > In-Person Payments > Manage Card Reader and connect to a reader
5. After the reader connects, it should automatically start the update 
6. Tapping the cancel button should cancel the update


https://user-images.githubusercontent.com/8739/134147854-12f5e238-90d0-4f74-b4a1-1a0a71852dac.mp4



Update release notes:
